### PR TITLE
Use range iterators for `Map` in most cases

### PR DESCRIPTION
**Pull request by AaronRecord**
See https://github.com/godotengine/godot/pull/50284
Tries to convert iterators to be `const` whenever possible.

*Bugsquad edit: Follow-up to https://github.com/godotengine/godot/pull/50511.*

<details>
<summary>Python script used, although I've made some manual fixup changes too.</summary>

```py
import os
import re


def const_only_usage() -> bool:
	assigned_to_regex = r"[^\s]+? [&*]?[_a-zA-Z][_a-zA-Z0-9]* = &?(Object::cast_to<.+?>\()?" + re.escape(iterator_name) + r"\.value(((\.|->).+?)+)?;"
	iterated_through_regex = r"\([^\s]+? &[_a-zA-Z][_a-zA-Z0-9]* : " + re.escape(iterator_name) + r"\.value\)"
	
	# Assigned to a constant value, must be constant.
	if re.search(f"const {assigned_to_regex}", slice) != None:
		return True
	if re.search(f"const {iterated_through_regex}", slice) != None:
		return True

	# Regexes to check.
	const_checks = [
		# Value or sub-value assigned.
		re.compile(iterator_name + r"\.value(((\.|->).+?)+)?(\[.+?\])? = .+?;"),
		# Assigned to non-const value.
		re.compile(assigned_to_regex),
		re.compile(iterated_through_regex),
		# Non-const method called on value or sub-value.
		re.compile(iterator_name + r".value(((\.|->).+?)+)?(->|\.)((push_packet|put_packet|commit|notify_tile_data_properties_should_change|_update_doc|clear|normalize|setup_local_to_scene|remove)\(|(write)\[)"),
		# Value passed as a non-const argument.
		re.compile(r"(_update_context|_set_subclass_path)\(" + re.escape(iterator_name) + r".value"),
		re.compile(r"(to_erase\.push_back|dirty_quadrant_list\.add)\(&?" + iterator_name)
	]

	for const_check in const_checks:
		if const_check.search(slice) != None:
			return False
	return True


directory = "C:/Users/Aaron/Projects/GodotSource/godot"
types = ("cpp", "h", "hpp")

os.chdir(directory)

# Get a list of files to modify.
grabbed_files = []
for root, dirs, files in os.walk(directory, topdown=True):
	dirs[:] = [d for d in dirs if d not in (".git", "thirdparty")]
	for file in files:
		if not file.endswith(types) or ".gen" in file:
			continue

		grabbed_files.append(os.path.join(root, file))


map_iter_re = re.compile(r"for ?\((const )?Map<(.+?), (.+?)>::Element \*(.+?) = (.+?)\.front\(\); \4; \4 = \4->next\(\)\) \{")
for grabbed_file in grabbed_files:
	# Read in the file.
	with open(grabbed_file, 'r', encoding="utf-8", newline="\n") as file:
		file_data = file.read()

	pos = 0
	while True:
		result = map_iter_re.search(file_data, pos)
		if not result:
			# Next file.
			break
		
		pos = result.end(0)

		is_const = result[1] != None
		key_type = result[2]
		value_type = result[3]
		iterator_name = result[4]
		container_name = result[5]

		# Discard custom comparators.
		if value_type.count(",") > value_type.count(">"):
			value_type = value_type[:value_type.rfind(",")]

		# Workaround for weird cases.
		if container_name in ("state.MeshNodes", "physics_2d_mouseover", "physics_2d_shape_mouseover"):
			# Should state.MeshNodes be deleted? It's an empty loop.
			continue

		# Get closing bracket position.
		bracket_count = 1
		bracket_pos = result.end(0)
		while bracket_count > 0:
			if file_data[bracket_pos] == "{":
				bracket_count += 1
			elif file_data[bracket_pos] == "}":
				bracket_count -= 1
			
			bracket_pos += 1
	
		slice = file_data[result.end(0):bracket_pos]
		if re.search(r"\b" + re.escape(iterator_name) + r"->next\(\)", slice):
			# Skip iterators that make use of `->next()`.
			continue
		if re.search(r"\b" + re.escape(container_name) + r"(->|\.)erase\(", slice):
			# Container is modified, I'm not sure how this is handeled.
			continue
		if f"_make_quadrant_dirty({iterator_name})" in slice:
			# See tile_map.cpp
			continue
		if "Object::cast_to<VisualShaderNodeGlobalExpression>(E->get().node.ptr())" in slice:
			# Not sure how to handle.
			continue

		replace_with = {"get": "value", "value": "value", "key": "key"}
		for key, value in replace_with.items():
			slice = re.sub(
				r"(?<=\b)" + re.escape(iterator_name) + r"->" + re.escape(key) + r"\(\)",
				iterator_name + "." + value,
				slice
			)
		# Convert iterators that weren't previously const to const if possible.
		if not is_const:
			is_const = const_only_usage()

		file_data = file_data[:result.end(0)] + slice + file_data[bracket_pos:]

		slice = file_data[result.start(0):result.end(0)]
		const = "const " if is_const else ""
		slice = f"for ({const}KeyValue<{key_type}, {value_type}> &{iterator_name} : {container_name}) {{"
		file_data = file_data[:result.start(0)] + slice + file_data[result.end(0):]

	# Write the file out again.
	with open(grabbed_file, 'w', encoding="utf-8", newline="\n") as file:
		file.write(file_data)

```
</details>

### CODE REVIEW

**Comment by LightningAA** [comment]
Passing checks 😄
I tried to get as cases as possible with my script, but there's only so much you can do without actually parsing the C++, so I ended up fixing up the last couple cases manually. I skipped the cases that did something peculiar, but this should cover the vast majority of cases.

(sorry for all the `force-pushes`, initially I was hesistant to try building locally since I usually build for `3.x` since the `4.0` binaries don't run on my computer (and I wanted to avoid having to rebuild for 3.x later), but in hindsight that would have been a lot smarter)

**Comment by LightningAA** [comment]
Rebased

**Comment by LightningAA** [comment]
(rebased)

**Comment by LightningAA** [comment]
Rebased. Ideally it'd be nice to get this merged sooner rather than later so it doesn't keep having to be rebased, but I'll be patient as necessary.

**Comment by akien-mga** [comment]
I think now would be a good time to merge this, so if you can rebase it again I'll make sure it's fast-tracked.

**Comment by LightningAA** [comment]
@akien-mga rebased :)

File: core/config/project_settings.cpp
Comment (akien-mga) [inline core/config/project_settings.cpp:791]: Looks like this one was missed?

**Review by akien-mga — APPROVED** [review]
Looks great to me. I only reviewed the changes in `core` in-depth, but I assume the rest is similar. We'll see once merged if there was any niche case where the conversion would have changed behavior.

**Comment by LightningAA** [comment]
> For the reference, I found a few occurrences of for (Map not converted yet in this PR, I don't know if it's intended or an oversight

It was intentional, #50284 says
> If Set/List/Map container iteration is done in reverse, or does something strange, leave as is

The ones I left would require more changes before switching to range iterators, but I wanted to limit this PR to the straightforward cases to keep it easier to review, and then a follow up PR could be done to get the rest of the cases. It could also be done in this PR as a separate commit, but I prefer smaller PRs because they are easier to review and merge.

File: core/config/project_settings.cpp
Comment (LightningAA) [reply core/config/project_settings.cpp:791]: No, it was intentional, because of `E != props.front()`

**Comment by LightningAA** [comment]
```
R    s
 e  a  e
  b     d
```

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
